### PR TITLE
Improve workload generator

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please describe this PR:**

This is to make wds easy to understand.

### CODE REVIEW

File: pilot/pkg/xds/workload.go [hunk 1]
```diff
@@ -51,22 +51,23 @@ func (e WorkloadGenerator) GenerateDeltas(
...
-	if !w.Wildcard {
-		// We only need this for on-demand. This allows us to subscribe the client to resources they
-		// didn't explicitly request.
-		// For wildcard, they subscribe to everything already.
-		// TODO: optimize me
-		additional := e.Server.Env.ServiceDiscovery.AdditionalPodSubscriptions(proxy, addresses, subs)
-		addresses.Merge(additional)
+	var addresses sets.String
+	if isReq {
+		// this is from request, we only send response for the subscribed address
```
Comment (costinm) [inline pilot/pkg/xds/workload.go:58]: To make sure I understand: if the XDS request subscribes to "A", we send response only for "A". If a following XDS request subscribes to "B" - is the intent to send only the response to B (not A) ? 

And I assume if A has changed - an independent response will be sent ? 

May be worth adding few more lines of comment.

**Review by costinm — COMMENTED** [review]
LGTM. Are all other delta resources behaving this way ?

**Comment by hzxuzhonghu** [comment]
>LGTM. Are all other delta resources behaving this way?

@costinm The other resources behave differently, they are simple. While workload is a very complicated resource. We support three kinds of resource names and AdditionalPodSubscriptions makes it even harder to understand.

File: pilot/pkg/xds/workload.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/xds/workload.go:58]: >To make sure I understand: if the XDS request subscribes to "A", we send response only for "A". If a following XDS request subscribes to "B" - is the intent to send only the response to B (not A) ?

Yes

If A changed, we will trigger push differently behaves like endpoint update

File: pilot/pkg/xds/workload.go [hunk 2]
```diff
@@ -51,20 +51,28 @@ func (e WorkloadGenerator) GenerateDeltas(
...
-	addresses := updatedAddresses
-	// If it is not a wildcard, filter out resources we are not subscribed to
-	if !w.Wildcard {
-		addresses = addresses.Intersection(subs)
+	var addresses sets.String
+	if isReq {
+		// this is from request, we only send response for the subscribed address
+		// At t0, a client request A, we only send A and aditional resources back to the client.
+		// At t1, a client request B, we only send B and aditional resources back to the client, no A here.
+		addresses = req.Delta.Subscribed
```
Comment (howardjohn) [inline pilot/pkg/xds/workload.go:61]: Seems like we are ignoring req.Delta.Unsubscribed?

Comment (hzxuzhonghu) [reply pilot/pkg/xds/workload.go:61]: We already removed the `Unsubscribed` when calculate `subs` in `deltaWatchedResources`.

So here we donot need to remove again

**Comment by hzxuzhonghu** [comment]
ping @howardjohn @costinm

**Review by howardjohn — APPROVED** [review]

**Comment by hzxuzhonghu** [comment]
Thank you john

**Pull request merged**